swig/python/conftest.py

```diff
@@ -0,0 +1,140 @@
+import ast
+import doctest
+import functools
+
+import numpy as np
+import pytest
+
+from osgeo import gdal, ogr, osr
+
+
+###
+# Make modules available to doctests
+#
+# Adding moduels to doctest_namespace allows us to avoid writing "from osgeo import gdal"
+# in every example.
+###
+@pytest.fixture(scope="session", autouse=True)
+def global_setup(doctest_namespace):
+    doctest_namespace["gdal"] = gdal
+    doctest_namespace["ogr"] = ogr
+    doctest_namespace["osr"] = osr
+    doctest_namespace["np"] = np
+    doctest_namespace["pytest"] = pytest
+
+
+###
+# Custom output checker
+#
+# doctest relies on an OutputChecker object to determine if test outputs
+# match their expected values. pytest has its own OutputChecker implementation
+# but does not provide any hooks for further customizations. So we inject a
+# custom checker directly into pytest's internals.
+###
+
+import _pytest.doctest
+
+base_checker = _pytest.doctest._get_checker()
+
+
+class GDALOutputChecker(doctest.OutputChecker):
+    @staticmethod
+    def _strip_comments(want):
+        lines = want.split("\n")
+        for i in range(len(lines)):
+            comment_pos = lines[i].find("#")
+            if comment_pos != -1:
+                lines[i] = lines[i][:comment_pos]
+        return "\n".join(lines)
+
+    @staticmethod
+    def _strip_warnings(want):
+        lines = want.split("\n")
+        return "\n".join(x for x in lines if not x.startswith("Warning"))
+
+    def check_output(self, want, got, flags):
+        want = self._strip_warnings(want)
+
+        # custom checking method for arrays
+        # pytest's built-in NUMBER flag is not suitable because
+        # it only allows differences to the right of the decimal
+        # point.
+        if "rtol:" in want:
+            pos = want.find("rtol:")
+            rtol = float(want[(pos + 5) :].strip())
+            want = np.array(ast.literal_eval(want))
+            got = np.array(ast.literal_eval(got))
+            return np.allclose(want, got, rtol=rtol)
+
+        if "# random" in want:
+            want = np.array(ast.literal_eval(want))
+            got = np.array(ast.literal_eval(got))
+            return np.array_equal(want.shape, got.shape)
+
+        if "# no-check" in want:
+            return True
+
+        want = self._strip_comments(want)
+
+        return base_checker.check_output(want, got, flags)
+
+
+_pytest.doctest.CHECKER_CLASS = GDALOutputChecker
+
+###
+# Path substitutions
+#
+# We want to write concise documentation using short filenames like gdal.Open("byte.tif")
+# Modifying the working directory before each doctest is tedious and error-prone.
+# Instead we store a list of abbreviated filenames and monkey-patch GDAL functions to
+# substitute the full path name before running the example.
+###
+
+files = {
+    "byte.tif": "autotest/gcore/data/byte.tif",
```

Comment:
There is an error on Windows about not finding the "autotest/..." files.  I believe that the current directory from which this is run must be the build directory. On Unix, there are symbolic links for convenience for $src_dir/autotest/{subdir} to $build_dir/autotest/{subdir}, but not on Windows.

So safer would be to have a mapping like

files = {
    "byte.tif":  abs_path_to("autotest/gcore/data/byte.tif"),
    ...
}

with abs_path_to that prepends the path of  __file__ / .. / ..